sql: support NULL in tuple comparison

madelynnblue opened the pull request:
Since parser.CmpOp.fn's signature returns a DBool, use a sentinel error
to indicate NULL. A cleaner solution may be to change the signature to
allow for explicitly returning a NULL.

Matches postgres behavior.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6370)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/eval.go, line 738 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6370#-KGSAkRUUllRIs-3l4QL:-KGSAkRVnYOTcGJxGH0S:888619368) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d5d384f49f5390fc9e17385017dc8b9cc09bb1/sql/parser/eval.go#L738)):*
Probably doesn't matter, but I preferred the prior version which had `cmpTuple` return an `int`. I realize that you'll need to return an `error` as well, but returning `1, cmpNull` seems workable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6370#-:-KGSBD54E0-tLYVxxCDP:1417088721)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/eval.go, line 738 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6370#-KGSAkRUUllRIs-3l4QL:-KGSqjdLgEvOnIzE23Fk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d5d384f49f5390fc9e17385017dc8b9cc09bb1/sql/parser/eval.go#L738)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6370)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: support NULL in tuple comparison

Since parser.CmpOp.fn's signature returns a DBool, use a sentinel error
to indicate NULL. A cleaner solution may be to change the signature to
allow for explicitly returning a NULL.

Matches postgres behavior.
